[Blog Post] Pulumi Recommended Practices: The Basics (was PR 652) (#707)
pulumi/pulumi-hugo: themes/default/content/blog/pulumi-recommended-patterns-the-basics/index.md

* [TypeScript](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html)
* Python: [PEP 8](https://www.python.org/dev/peps/pep-0008/), [PEP 20](https://www.python.org/dev/peps/pep-0020/), and [an additional blog post](https://data-flair.training/blogs/python-best-practices/)
* [C#](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions)
* Golang: [Golang docs](https://golangdocs.com/golang-best-practices) and [a talk](https://talks.golang.org/2013/bestpractices.slide#1)

## Pulumi recommended patterns

Now we're all set, so let's dive into writing a Pulumi application.

### Give your resources a name

The first thing you'll notice when you write a Pulumi app is that every single resource must have its own name. This *Pulumi resource name* is used inside your stack state file to uniquely identify your resources.

Depending on the resource type, a second name (a *cloud resource name* if you like) may also be provided. If this *cloud resource name* is unspecified, Pulumi uses the *Pulumi resource name* and appends a random string. The random string helps in avoiding name collisions between cloud resources.

> As a best practice, we recommend that you explicitly name all (or at least most) of your *cloud resources*. Being explicit with resource names allows you to use string manipulation more easily rather than having to work with an [Output]({{< relref "/docs/intro/concepts/inputs-outputs" >}}).

Reviewer: Maybe this is phrased wrong, but I'd say that the default and encouraged practice should be to never name the cloud resource and rely on the random name. When a cloud resource name is required to be predefined, then we also need to configure it with destroyBeforeCreate

Reviewer 2: Ok, I think you're coming from the angle of `deleteBeforeReplace` and the cascading effects of 1) more resources taken down, and 2) potentially longer downtime during an update because a resource is first deleted and the recreated. Am I understanding you correctly?

On resource names, Mikhail posted a nice answer on the topic on SO. Not having naming scheme adds way too many uncertainties when people look at their cloud console and see a bunch of similarly named resources. Having to reconcile the Cloud console and pulumi resources is probably not a fun game to play. For me at least, I find it hard to do any kind of string manipulation with an `Output<string>` without adding too much extra code.

That said, I think the article could make it clearer that it should be a team decision for creating such scheme and highlight the down side. But random names are still a problem IMHO.

* Plugins and extension support to augment the IDE capabilities (issues and pull request management, prettier code, themes and colors, pair programming...)

And just as with the cloud vendors, you should definitely consider the do's, don'ts, and recommended patterns for the programming language you selected. Following these best practices gives you safe guardrails and streamlines your developer experience across teams sharing the same programming language. Here's the main ones for a few languages:

* [TypeScript](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html)
* Python: [PEP 8](https://www.python.org/dev/peps/pep-0008/), [PEP 20](https://www.python.org/dev/peps/pep-0020/), and [an additional blog post](https://data-flair.training/blogs/python-best-practices/)
* [C#](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions)
* Golang: [Golang docs](https://golangdocs.com/golang-best-practices) and [a talk](https://talks.golang.org/2013/bestpractices.slide#1)

## Pulumi recommended patterns

Now we're all set, so let's dive into writing a Pulumi application.

### Give your resources a name

> The first thing you'll notice when you write a Pulumi app is that every single resource must have its own name. This *Pulumi resource name* or a logical resource name if you prefer, is used inside your stack state file to uniquely identify your resources. This is also what you will see at first when exploring your stack resources in the Pulumi web console.

Author: ```suggestion
The first thing you'll notice when you write a Pulumi app is that every single resource must have its own name. This *Pulumi resource name*, or a logical resource name if you prefer, is used inside your stack state file to uniquely identify your resources. This name is also what you will find at first when exploring your stack resources in the Pulumi web console.
```

> Depending on the resource type, a second name (a *cloud resource name* or a *physical resource name* if you like) may also be provided. This is the name that will be visible in your Cloud vendor web console. If this *cloud resource name* is unspecified, Pulumi uses the *Pulumi resource name* and appends a [random suffix]({{< relref "/docs/intro/concepts/resources/#autonaming" >}}) to it to make it unique.

Author: Adjustment to remove inaccessible wording. Also, one thing Hugo seems to require with `relref` for future reference: If you want to use an anchor link on another page, you have to remove the trailing slash. I forget exactly why, but something to file away for any future articles :)
```suggestion
Depending on the resource type, a second name (a *cloud resource name* or a *physical resource name* if you like) may also be provided. This name will be what is used in your cloud vendor's web console. If this *cloud resource name* is unspecified, Pulumi uses the *Pulumi resource name* and appends a [random suffix]({{< relref "/docs/intro/concepts/resources#autonaming" >}}) to it to make it unique.
```

> As a recommendation, you should use explicit names for all your *Pulumi resource names* and then let Pulumi determine what the *Cloud resource names* should be. Let's Look at the benefits:

Author: ```suggestion
As a recommendation, you should use explicit names for all your *Pulumi resource names* and then let Pulumi determine what the *cloud resource names* should be. Let's explore the benefits:
```

> * As vendors have different requirements, *Cloud resource names* have different lengths and constraints. Pulumi knows this and will make sure the name matches the vendor requirements.

Author: ```suggestion
* As vendors have different requirements, *cloud resource names* have different lengths and constraints. Pulumi knows this fact and will make sure the name matches the vendor requirements.
```

* The random suffix ensures that two stacks for the same project can be deployed without their resources colliding.
* Finally, the random suffix allows Pulumi to do **zero-downtime** resource updates. Due to the way some cloud providers work, certain updates require replacing resources rather than updating them in place. By default, Pulumi creates replacements first, then updates the existing references to them, and finally deletes the old resources.

![a code snippet that shows a Pulumi program that uses resource names and then manipulates that string in the program](images/resource-names.svg)

![an animation showing the Pulumi program running with output that gives the stated resource name and a resource name with an appended string](images/pulumi-up-resource-names.svg)

In this example, the resource `inputBucket` only has a Pulumi resource name set whereas `outputBucket` has both the Pulumi resource name and the cloud resource name set. As a result of the lack of a cloud resource name, `inputBucketName` uses the Pulumi resource name as the base name with a random suffix. As for `outputBucketName`, the cloud resource name is set as specified by our code.

> Along with this recommendation, using a unique prefix in the name of your resources is an elegant solution quickly identify resources across multiple projects and stacks. A way to achieve this is to use a customer name or a project name. As [suggested](https://stackoverflow.com/a/69270933) by my colleague Mikhail, the resource name prefix may be composed of multiple values, such as `${organization}-${tenant}-${environment}-${resourceName}`.

Author: ```suggestion
Along with this recommendation, using a unique prefix in the name of your resources is an elegant solution to identify resources across multiple projects and stacks quickly. A way to achieve this goal is to use a customer name or a project name. As [suggested](https://stackoverflow.com/a/69270933) by my colleague Mikhail, the resource name prefix may be composed of multiple values, such as `${organization}-${tenant}-${environment}-${resourceName}`.
```

> Here is a few links to better understand physical names and auto-naming:

Author: ```suggestion
Here are a few links to understand physical names and auto-naming a bit better:
```

> * [Resources and auto-naming]({{< relref "/docs/intro/concepts/resources/#autonaming" >}})

Author: ```suggestion
* [Resources and auto-naming]({{< relref "/docs/intro/concepts/resources#autonaming" >}})
```

> * [Infrastructure as Code Resource Naming]({{< relref "/blog/infrastructure-as-code-resource-naming/" >}})

Author: ```suggestion
* [Infrastructure as Code Resource Naming]({{< relref "/blog/infrastructure-as-code-resource-naming" >}})
```

> * [Why do resource names have random hex character suffixes?]({{< relref "/docs/troubleshooting/faq/#why-do-resource-names-have-random-hex-character-suffixes" >}})

Author: ```suggestion
* [Why do resource names have random hex character suffixes?]({{< relref "/docs/troubleshooting/faq#why-do-resource-names-have-random-hex-character-suffixes" >}})
```

Using Pulumi and its secret management capabilities is an important step toward more secure deployments in general. But also, it enables developers to reach "zero-knowledge" of any sensitive information, effectively separating code and deployment from passwords and other sensitive information.

## Summary

For this first article, we laid down the important foundations of infrastructure-as-code. Here is a quick summary:

* Follow your vendors' best practices as early as possible.
* Choose the programming language and the IDE you're the most comfortable with.
* Give your Pulumi resources a name in Pulumi and in your cloud environment.
* Leverage Pulumi [secrets]({{< relref "/docs/intro/concepts/secrets" >}}) to protect your sensitive information during deployment.

In the next article, we'll discuss code reusability and code documentation to accelerate your developer experience. We show you how to have a multi-stack environment, protect critical resources, and more.

## Acknowledgement

Thank you to Stephanie Swenson at [Furō](https://www.furo.io/) for reviewing and suggesting many great edits in this article.

> Thank you to Laura Santamaria, David Flanagan, Joe Duffy and Luke Hoban at [Pulumi](https://www.pulumi.com/) for sharing their knowledge, reviewing, and suggesting edits in this article.

Author: ```suggestion
Thank you to Laura Santamaria, David Flanagan, Joe Duffy, and Luke Hoban at [Pulumi](https://www.pulumi.com/) for sharing their knowledge, reviewing, and suggesting edits in this article.
```
